src/main/java/me/moodcat/api/SongAPI.java

```diff
@@ -48,4 +71,131 @@ public Song getSongById(@PathParam("id") final int id) {
+            final ClassificationRequest classification) throws InvalidClassificationException {
+        final Song song = this.songDAO.findById(id);
+
+        assertDimensionIsValid(classification.getValence());
+        assertDimensionIsValid(classification.getArousal());
+
+        if (song.getNumberOfPositiveVotes() < MINIMUM_NUMBER_OF_POSITIVE_VOTES) {
+            song.setValenceArousal(adjustSongVector(classification, song));
+            this.songDAO.merge(song);
+        }
+
+        return classification;
+    }
+
+    /**
+     * Process a vote to a song. A vote is either "like" or "dislike".
+     *
+     * @param soundCloudId
+     *            The soundCloudID.
+     * @param vote
+     *            The vote.
+     * @return The song object, if the process was succesful.
+     * @throws InvalidVoteException
+     *             Thrown if an invalid vote was cast.
+     */
+    @POST
+    @Path("{id}/vote/{vote}")
+    @Transactional
+    public Song voteSong(@PathParam("id") final int soundCloudId,
+            @PathParam("vote") final String vote) throws InvalidVoteException {
+        final Song song = this.songDAO.findBySoundCloudId(soundCloudId);
+
+        if (vote.equals("like")) {
+            song.increaseNumberOfPositiveVotes();
+        } else if (vote.equals("dislike")) {
+            song.decreaseNumberOfPositiveVotes();
+        } else {
+            throw new InvalidVoteException();
+        }
+
+        return song;
+    }
+
+    private void assertDimensionIsValid(final double value) throws InvalidClassificationException {
+        if (!Arrays.asList(ACCEPTED_DIMENSION_VALUES).contains(value)) {
+            throw new InvalidClassificationException();
+        }
+    }
+
+    private VAVector adjustSongVector(final ClassificationRequest classification, final Song song) {
+        final VAVector classificationVector = new VAVector(
+
+                classification.getValence(),
+                classification.getArousal());
+
+        final VAVector songVector = song.getValenceArousal();
+
+        final VAVector scaledDistance = classificationVector.subtract(songVector)
+                .multiply(CLASSIFICATION_WEIGHT);
+
+        final VAVector updatedVector = songVector.add(scaledDistance);
+        return updatedVector;
+    }
+
+    /**
+     * Classificationrequest with the arousal and valence the user would classify the specific song
+     * for.
+     *
+     * @author JeremybellEU
+     */
+    @AllArgsConstructor
+    protected static class ClassificationRequest {
+
+        /**
+         * The valence for the song.
+         */
+        @Getter
+        private final double valence;
+
+        /**
+         * The arousal for the song.
+         */
+        @Getter
+        private final double arousal;
+    }
+
+    /**
+     * Thrown if the classification was invalid.
+     *
+     * @author JeremybellEU
+     */
+    protected static class InvalidClassificationException extends Exception {
```

Comment:
Please either use `IllegalArgumentException` or extend `IllegalArgumentException` for consistency.